Kitodo-DataManagement/src/main/java/org/kitodo/data/database/beans/IndexingKeyworder.java

```diff
@@ -0,0 +1,502 @@
+    private static final Pattern METADATA_SECTIONS_PATTERN =
+            Pattern.compile("<mets:dmdSec.*?o>(.*?)</kitodo:k", Pattern.DOTALL);
+    private static final Pattern RULESET_KEY_PATTERN =
+            Pattern.compile("key id=\"([^\"]+)\">(.*?)</key>", Pattern.DOTALL);
+    private static final Pattern RULESET_LABEL_PATTERN = Pattern.compile("<label[^>]*>([^<]+)", Pattern.DOTALL);
+
+    private static final Map<String, Map<String, Collection<String>>> rulesetCache = new HashMap<>();
+
+    private Set<String> titleKeywords = Collections.emptySet();
+    private Set<String> projectKeywords = Collections.emptySet();
+    private Set<String> batchKeywords = Collections.emptySet();
+    private Set<String> taskKeywords = Collections.emptySet();
+    private Set<String> taskPseudoKeywords = Collections.emptySet();
+    private Set<String> metadataKeywords = Collections.emptySet();
+    private Set<String> metadataPseudoKeywords = Collections.emptySet();
+    private String processId = null;
+    private Set<String> commentKeywords = Collections.emptySet();
+
+    public IndexingKeyworder(Process process) {
+        this.titleKeywords = filterMinLength(initTitleKeywords(process.getTitle()));
+        String projectTitle = Objects.nonNull(process.getProject()) ? process.getProject().getTitle() : "";
+        this.projectKeywords = filterMinLength(initSimpleKeywords(projectTitle));
+        this.batchKeywords = filterMinLength(initBatchKeywords(process.getBatches()));
+        var taskKeywords = initTaskKeywords(process.getTasksUnmodified());
+        this.taskKeywords = filterMinLength(taskKeywords.getLeft());
+        this.taskPseudoKeywords = filterMinLength(taskKeywords.getRight());
+        String place = null;
+        if (logger.isDebugEnabled()) {
+            place = "process " + process.getId() + " \"" + process.getTitle() + "\"";
+        }
+        var metadataKeywords = initMetadataKeywords(process, place);
+        this.metadataKeywords = filterMinLength(metadataKeywords.getLeft());
+        this.metadataPseudoKeywords = filterMinLength(metadataKeywords.getRight());
+        this.processId = process.getId().toString();
+        this.commentKeywords = filterMinLength(initCommentKeywords(process.getComments()));
+        if (logger.isTraceEnabled()) {
+            traceLogKeywords("indexterms.log");
+        }
+    }
+
+    /**
+     * Creates search terms for process titles. To do this, the process title is
+     * separated into groups of consecutive characters and numbers, and these
+     * are generated in their entirety and starting from the front or from the
+     * back. The latter is more common, as the last four digits of a PPN are
+     * used for the search and the hit rate is excellent.
+     * 
+     * @param processTitle
+     *            the title of the process
+     * @return keywords
+     */
+    private static Set<String> initTitleKeywords(String processTitle) {
```

Comment:
Is this method and similar methods are really needed? Should this not be done by ElasticSearch/Opensearch in a internal way without bothering every one who is using it? This will increase the execution time and the data size on ElasticSearch/OpenSearch server without knowing that a internal handling is better.

Maybe the submitted data must provided in an other way to get this handled by ElasticSearch/Opensearch itself instead of implementing and maintaining it on our side?

Reply:
I think it's possible that you can build this into OpenSearch, but then you'll have to compile your own OpenSearch binaries and you won't be able to just install OpenSearch from the distribution. Maybe that makes sense if we have several of them, but for administrator convenience, it would be more convenient to install the web application and have it just work.

Reply:
As @solth already written: I thought that this could be configured on the ES / OS side or through the type of data so no change is needed explicit on ES / OS side itself instead of implementing this on our side. Maybe my imagination is wrong but this would blow up the data size of the ES / OS index a lot if you have many processes and big meta data (task data, comment data, data from the meta data files itself, ...) entries per process. But we will see how the handling and outcome is.

Reply:
I'm sorry I have not the knowledge to do this as this is not my part of my work. I should look into the changes and I should comment them if I'm unsure about the changes. I don't know which exact requirements are made as I'm not involved in this discussions. The solutions looks complex and bad to maintain and I still thinking that this is not really necessary to do this required work every time a ES / OS should be used. Maybe this is limited because of the way of hibernate-search or through other limitations. Maybe I'm overrating what should be done by ES / OS on default and this stuff must be done by everyone how is using ES / OS. Maybe using the [binding and brigdes](https://docs.jboss.org/hibernate/search/6.2/reference/en-US/html_single/index.html#binding) could be a start of writing self defined bridges in a more clean way if they are a possible solution.

Reply:
I might need to revisit the requirements, but from my understanding, the main source of complexity seems to be supporting multi-term searches like "digi abc" that would match titles such as digi_1234_abc, ignoring separators like underscores or spaces. This feels like overkill and might go beyond what’s necessary.

A simpler approach could be focusing on searches based on three or more consecutive characters/tokens, which would still allow matching relevant titles.

For example, searching for "abc" or "1234" should match the title digi_1234_abc if the search logic considers tokens generated by an ngram tokenizer with min_gram: 3. This would generate tokens such as:


"digi_1234_abc" → ["dig", "digi", "igi", "gi_", "i_1", "123", "234", "abc"]
With this setup:

- A search for "abc" would match because "abc" exists as a token.
- A search for "igi" would match because "igi" is also generated as a token.

Reply:
Similar to the programming way in Java, it is also possible to define the same in [custom index settings](https://docs.jboss.org/hibernate/stable/search/reference/en-US/html_single/#backend-elasticsearch-configuration-index-settings) which could even used for other general or index specific adjustments.